### PR TITLE
Small performance increase for std.conv.text/wtext/dtext

### PR DESCRIPTION
**Pull request by JackStouffer**
Saves about 5 - 7 hnsecs.

### CODE REVIEW

File: std/conv.d [hunk 1]
```diff
@@ -3711,10 +3711,24 @@ private S textImpl(S, U...)(U args)
...
+        if (__ctfe)
+        {
+            auto result = to!S(args[0]);
+            foreach (arg; args[1 .. $])
+                result ~= to!S(arg);
+            return result;
+        }
+        else
+        {
+            auto app = appender!S();
```
Comment (DmitryOlshansky) [inline std/conv.d]: if args.length == 1 this will be slower I guess

**Comment by JackStouffer** [comment]
@DmitryOlshansky Fixed

**Comment by DmitryOlshansky** [comment]
> Saves about 5 - 7 hnsecs.

How do you measure it?

**Comment by JackStouffer** [comment]
@DmitryOlshansky 

```
import std.stdio;
import std.datetime;
import std.conv;

enum testCount = 1_000_000;

S textImpl(S, U...)(U args)
{
    static if (U.length == 0)
    {
        return null;
    }
    else
    {
        auto result = to!S(args[0]);
        foreach (arg; args[1 .. $])
            result ~= to!S(arg);
        return result;
    }
}

S textImpl2(S, U...)(U args)
{
    static if (U.length == 0)
    {
        return null;
    }
    else
    {
        import std.array : appender;

        if (__ctfe)
        {
            auto result = to!S(args[0]);
            foreach (arg; args[1 .. $])
                result ~= to!S(arg);
            return result;
        }
        else
        {
            auto app = appender!S();
            app.reserve(U.length * S.sizeof);

            foreach (arg; args)
                app.put(to!S(arg));
            return app.data;
        }
    }
}

void main()
{
    auto result = to!Duration(benchmark!(() => textImpl!dstring(42, ' ', 1.5, ": xyz", 333, "Another test"))(testCount)[0] / testCount);
    auto result2 = to!Duration(benchmark!(() => textImpl2!dstring(42, ' ', 1.5, ": xyz", 333, "Another test"))(testCount)[0] / testCount);

    writeln("result", "\t", result);
    writeln("result2", "\t", result2);
}
```

**Comment by JackStouffer** [comment]
I really don't see why the auto tester is failing.

File: std/conv.d [hunk 2]
```diff
@@ -3709,12 +3709,30 @@ private S textImpl(S, U...)(U args)
...
+            app.reserve(U.length * S.sizeof);
```
Comment (aG0aep6G) [inline std/conv.d]: Why `S.sizeof`? As far as I see, it's just an arbitrary number, determined by rule of thumb, and using `S.sizeof` for this suggests a relation where there is none.

File: std/conv.d [hunk 3]
```diff
@@ -3709,12 +3709,30 @@ private S textImpl(S, U...)(U args)
...
+
+            foreach (arg; args)
+                app.put(to!S(arg));
```
Comment (aG0aep6G) [inline std/conv.d]: Still creating one string per argument. How about using `std.format.formatValue` to format directly into the appender?

**Comment by aG0aep6G** [comment]
Are you sure it's `Appender` that fails CTFE, and not `to!S`?

**Comment by aG0aep6G** [comment]
> I really don't see why the auto tester is failing.

Compiler bug. Filed it here: https://issues.dlang.org/show_bug.cgi?id=15992

**Comment by JackStouffer** [comment]
>Why S.sizeof? 

|

> Are you sure it's Appender that fails CTFE, and not to!S?

Yeah, not sure what I was thinking there.

> Still creating one string per argument. How about using std.format.formatValue to format directly into the appender?

It's not `@safe`, and it breaks a bunch of test in std.datetime.

**Comment by aG0aep6G** [comment]
> It's not @safe, and therefore it breaks a bunch of test in std.datetime. Also, I don't know if it's a good idea to make this unsafe in general.

I see. Yeah, no, let's keep things `@safe`.

File: std/conv.d (hunk 3 above)
Comment (9il) [inline std/conv.d]: +1, please use formatValue

File: std/conv.d [hunk 4]
```diff
@@ -3709,12 +3709,30 @@ private S textImpl(S, U...)(U args)
...
+    }
     else
     {
-        auto result = to!S(args[0]);
-        foreach (arg; args[1 .. $])
-            result ~= to!S(arg);
-        return result;
+        import std.array : appender;
+
+        if (__ctfe)
```
Comment (9il) [inline std/conv.d]: appender should work in ctfe thought

**Comment by JackStouffer** [comment]
@aG0aep6G @9il fixed

**Comment by JackStouffer** [comment]
Can someone add the "blocked" tag to this?

**Comment by schveiguy** [comment]
One heuristic optimization -- until an argument produces some actual string output, there's no need to fire up an appender. Essentially, you have the single-argument case, no need to copy the string that was just generated and throw that string away. Not sure if it helps or hurts the performance.

**Comment by schveiguy** [comment]
See my comment here: https://github.com/dlang/phobos/pull/4460#issuecomment-228350769

**Comment by JackStouffer** [comment]
Can someone kick the auto-tester?

File: std/conv.d [hunk 5]
```diff
@@ -3709,12 +3709,19 @@ private S textImpl(S, U...)(U args)
...
-        auto result = to!S(args[0]);
-        foreach (arg; args[1 .. $])
-            result ~= to!S(arg);
-        return result;
+        import std.array : appender;
+
+        auto app = appender!S();
+
+        foreach (arg; args)
+            app.put(to!S(arg));
```
Comment (9il) [inline std/conv.d]: `app.formattedWrite(fmt, arg);` should be used instead.

**Comment by 9il** [comment]
You can add `trusted`, but please use `formatValue` instead

**Comment by JackStouffer** [comment]
> You can add trusted

What if one of the arguments has a memory unsafe `toString`?

**Comment by JackStouffer** [comment]
Ready to go

**Comment by 9il** [comment]
> You can add trusted
> What if one of the arguments has a memory unsafe toString?

Implementations should not be aware of what kind of `toString` is used, if they have both.

**Comment by 9il** [comment]
Is current implementation is always @safe?

**Comment by JackStouffer** [comment]
> Is current implementation always `@safe`?

Depends on the implementation of `to` used for each of the arguments.

**Comment by JackStouffer** [comment]
Ok, I think I reached a solution. In general, marking templates `@trusted` is a very bad idea. But, for built-in types, we can safely assume that conversions to strings are `@safe`, so we can safely mark a function `@trusted` that operates on only built-in types. For all other types, we use an inferred template which will be safe or not depending on the implementation of `to` chosen.

This code provides a 2x speed up when using built in types

```
$ dmd -O -inline -release test.d && ./test
result	2 μs and 8 hnsecs  # old
result2	1 μs and 4 hnsecs  # new
result3	2 μs and 8 hnsecs  # old
result4	2 μs and 2 hnsecs  # new

$ ldc2 -O5 -release test.d && ./test
result	2 μs  # old
result2	1 μs  # new
result3	2 μs  # old
result4	1 μs and 8 hnsecs  # new
```

Here's the code

```
import std.stdio;
import std.datetime;
import std.conv;
import std.meta;
import std.traits;

enum testCount = 1_000_000;

S textImpl(S, U...)(U args)
{
    static if (U.length == 0)
    {
        return null;
    }
    else
    {
        auto result = to!S(args[0]);
        foreach (arg; args[1 .. $])
            result ~= to!S(arg);
        return result;
    }
}

private S textImpl2(S, U...)(U args) if (U.length <= 1)
{
    static if (U.length == 0)
    {
        return null;
    }
    else static if (U.length == 1)
    {
        return to!S(args[0]);
    }
}

// @trusted version that's faster, needs to be @trusted due to formattedWrite
// excludes non-built-in types as we cannot trust the toString
// of user/library defined types to be memory safe
private S textImpl2(S, U...)(U args) @trusted
    if (U.length > 1 && allSatisfy!(isBuiltinType, U))
{
    import std.array : appender;
    import std.format : singleSpec, formatValue;

    auto app = appender!S();
    auto spec = singleSpec("%s");

    foreach (arg; args)
        formatValue(app, arg, spec);

    return app.data;
}

// This version will infer safety or not based on the chosen
// implementation of std.conv.to, slower than the other overload
private S textImpl2(S, U...)(U args)
    if (U.length > 1 && !allSatisfy!(isBuiltinType, U))
{
    import std.array : appender;

    auto app = appender!S();

    foreach (arg; args)
        app.put(to!S(arg));

    return app.data;
}

struct Type
{
    int a;

    string toString()
    {
        return "Another test";
    }
}

void main()
{
    auto t = Type();

    auto result = to!Duration(benchmark!(() => textImpl!dstring(42, ' ', 1.5, ": xyz", 333, "Another test"))(testCount)[0] / testCount);
    auto result2 = to!Duration(benchmark!(() => textImpl2!dstring(42, ' ', 1.5, ": xyz", 333, "Another test"))(testCount)[0] / testCount);

    auto result3 = to!Duration(benchmark!(() => textImpl!dstring(42, ' ', 1.5, ": xyz", 333, t))(testCount)[0] / testCount);
    auto result4 = to!Duration(benchmark!(() => textImpl2!dstring(42, ' ', 1.5, ": xyz", 333, t))(testCount)[0] / testCount);

    writeln("result", "\t", result);
    writeln("result2", "\t", result2);
    writeln("result3", "\t", result3);
    writeln("result4", "\t", result4);
}
```

**Comment by JackStouffer** [comment]
formatValue is also impure, so that's a no go

**Comment by JackStouffer** [comment]
Ping @schveiguy @DmitryOlshansky @9il

**Comment by DmitryOlshansky** [comment]
LGTM

**Comment by schveiguy** [comment]
Auto-merge toggled on

**Comment by schveiguy** [comment]
I'd like to have seen an attempt to avoid appender until necessary, but OK.

**Pull request merged**